### PR TITLE
Leveldb schema/key change for indexing

### PR DESCRIPTION
**Pull request by wu-hui**

### CODE REVIEW

File: Firestore/core/src/local/leveldb_key.cc [hunk 1]
```diff
@@ -115,6 +118,11 @@ enum ComponentLabel {
   /** A component containing the name of a named query. */
   QueryName = 18,
 
+  /** A component containing a index id. */
+  IndexId = 19,
+  IndexArrayValue = 20,
```
Comment (schmidt-sebastian) [inline Firestore/core/src/local/leveldb_key.cc:123]: Can you also add comments for IndexArrayValue/IndexDirectionalValue?

File: Firestore/core/src/local/leveldb_key.cc [hunk 2]
```diff
@@ -567,6 +587,21 @@ std::string Reader::Describe() {
       if (ok_) {
         absl::StrAppend(&description, " query_name=", query_name);
       }
+    } else if (label == ComponentLabel::IndexId) {
+      int32_t index_id = ReadIndexId();
+      if (ok_) {
+        absl::StrAppend(&description, " index_id=", index_id);
+      }
+    } else if (label == ComponentLabel::IndexArrayValue) {
+      const std::string& value = ReadIndexArrayValue();
```
Comment (schmidt-sebastian) [inline Firestore/core/src/local/leveldb_key.cc:596]: Don't store temporary in a const reference.

File: Firestore/core/src/local/leveldb_key.cc [hunk 3]
```diff
@@ -567,6 +587,21 @@ std::string Reader::Describe() {
...
+      if (ok_) {
+        absl::StrAppend(&description, " array_value=", value);
+      }
+    } else if (label == ComponentLabel::IndexDirectionalValue) {
+      const std::string& value = ReadIndexDirectionalValue();
```
Comment (schmidt-sebastian) [inline Firestore/core/src/local/leveldb_key.cc:601]: Don't store temporary in a const reference.

File: Firestore/core/test/unit/local/leveldb_key_test.cc [hunk 4]
```diff
@@ -490,6 +490,154 @@ TEST(NamedQueryKeyTest, Description) {
...
+  ASSERT_TRUE(absl::StartsWith(
+      LevelDbIndexEntryKey::Key(0, "user_id", "array_value_encoded",
+                                "directional_value_encoded", "document_id_99"),
+      table_key));
+
+  ASSERT_FALSE(absl::StartsWith(LevelDbIndexEntryKey::Key(0, "", "", "", ""),
+                                LevelDbIndexEntryKey::Key(1, "", "", "", "")));
+}
+
+struct IndexEntry {
```
Comment (schmidt-sebastian) [inline Firestore/core/test/unit/local/leveldb_key_test.cc:577]: Not sure about this, but I would move this to the top of the file. You could also just store LevelDbIndexEntryKey directly.

**Review by schmidt-sebastian — APPROVED** [review]

File: Firestore/core/src/local/leveldb_key.cc (hunk 2 above)
Comment (wu-hui) [reply Firestore/core/src/local/leveldb_key.cc]: Done.

File: Firestore/core/src/local/leveldb_key.h [hunk 5]
```diff
@@ -713,6 +730,150 @@ class LevelDbNamedQueryKey {
...
+  int32_t index_id() const {
+    return index_id_;
+  }
+
+  /** The user id for this entry. */
+  const std::string& user_id() const {
+    return user_id_;
+  }
+
+  /** The array index value encoded for this entry. */
```
Comment (schmidt-sebastian) [inline Firestore/core/src/local/leveldb_key.h:854]:
```suggestion
  /** The encoded array index value for this entry. */
```

File: Firestore/core/src/local/leveldb_key.h [hunk 6]
```diff
@@ -713,6 +730,150 @@ class LevelDbNamedQueryKey {
...
+  const std::string& array_value() const {
+    return array_value_;
+  }
+
+  /** The directional index value encoded for this entry. */
```
Comment (schmidt-sebastian) [inline Firestore/core/src/local/leveldb_key.h:859]:
```suggestion
  /** The encoded directional index value for this entry. */
```

File: Firestore/core/src/local/leveldb_key.h [hunk 7]
```diff
@@ -713,6 +730,150 @@ class LevelDbNamedQueryKey {
...
+  const std::string& directional_value() const {
+    return directional_value_;
+  }
+
+  /** The document name this entry points to. */
+  const std::string& document_name() const {
```
Comment (schmidt-sebastian) [inline Firestore/core/src/local/leveldb_key.h:865]: FYI https://github.com/firebase/firebase-android-sdk/pull/3383

**Pull request merged**